internal/initializer/tls.go

```diff
@@ -0,0 +1,332 @@
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+)
+
+const (
+	errGenerateCA              = "cannot generate CA certificate"
+	errParseCACertificate      = "cannot parse CA certificate"
+	errParseCAKey              = "cannot parse CA key"
+	errLoadOrGenerateSigner    = "cannot load or generate certificate signer"
+	errDecodeKey               = "cannot decode key"
+	errDecodeCert              = "cannot decode cert"
+	errFmtGetTLSSecret         = "cannot get TLS secret: %s"
+	errFmtCannotCreateOrUpdate = "cannot create or update secret: %s"
+)
+
+const (
+	// RootCACertSecretName is the name of the secret that will store CA certificates and rest of the
+	// certificates created per entities will be signed by this CA
+	RootCACertSecretName = "root-ca-certs"
+)
+
+// TLSCertificateGenerator is an initializer step that will find the given secret
+// and fill its tls.crt, tls.key and ca.crt fields to be used for External Secret
+// Store plugins
+type TLSCertificateGenerator struct {
+	namespace           string
+	caSecretName        string
+	tlsServerSecretName string
+	tlsClientSecretName string
+	subject             string
+	owner               []metav1.OwnerReference
+	certificate         CertificateGenerator
+	log                 logging.Logger
+}
+
+// TLSCertificateGeneratorOption is used to configure TLSCertificateGenerator behavior.
+type TLSCertificateGeneratorOption func(*TLSCertificateGenerator)
+
+// TLSCertificateGeneratorWithLogger returns an TLSCertificateGeneratorOption that configures logger
+func TLSCertificateGeneratorWithLogger(log logging.Logger) TLSCertificateGeneratorOption {
+	return func(g *TLSCertificateGenerator) {
+		g.log = log
+	}
+}
+
+// NewTLSCertificateGenerator returns a new TLSCertificateGenerator.
+func NewTLSCertificateGenerator(ns, caSecret, tlsServerSecret, tlsClientSecret, subject string, owner []metav1.OwnerReference, opts ...TLSCertificateGeneratorOption) *TLSCertificateGenerator {
+	e := &TLSCertificateGenerator{
+		namespace:           ns,
+		caSecretName:        caSecret,
+		tlsServerSecretName: tlsServerSecret,
+		tlsClientSecretName: tlsClientSecret,
+		subject:             subject,
+		owner:               owner,
+		certificate:         NewCertGenerator(),
+		log:                 logging.NewNopLogger(),
+	}
+
+	for _, f := range opts {
+		f(e)
+	}
+	return e
+}
+
+func (e *TLSCertificateGenerator) loadOrGenerateCA(ctx context.Context, kube client.Client, nn types.NamespacedName) (*CertificateSigner, error) {
+	caSecret := &corev1.Secret{}
+
+	err := kube.Get(ctx, nn, caSecret)
+	if resource.IgnoreNotFound(err) != nil {
+		return nil, errors.Wrapf(err, errFmtGetTLSSecret, nn.Name)
+	}
+
+	if err == nil {
+		kd := caSecret.Data[SecretKeyTLSKey]
+		cd := caSecret.Data[SecretKeyTLSCert]
+		if len(kd) != 0 && len(cd) != 0 {
+			e.log.Info("TLS CA secret is complete.")
+			return parseCertificateSigner(kd, cd)
+		}
+	}
+	e.log.Info("TLS CA secret is empty or not complete, generating a new CA...")
```

Comment:
Does "empty" here mean "not found"? Or are they two different cases?

Reply:
Since the secret is created without any data via helm, it means "empty" here. We're checking the secret content's completeness to not overwrite the certificates each time the crossplane pod restarts.

Reply:
Got it. I noticed we don't seem to be handling the case where the whole secret is actually not found above. I assume that's intentional though, since we'll just create one if it's not there.

Reply:
If the secret belongs to a provider or a function, we will create it. If it belongs to Crossplane, it should be created by Helm and if it's deleted afterward, the Crossplane pod will complain about not being able to mount it. In that case, it should be created again manually.